[#2] 회원가입 기능 구현 PR

msugo1 opened the pull request:
## [#2] 피드백 반영

    * Controller 계층
    - URL관련 웹 API 디자인 지침 반영(/idCheck -> /idcheck)
    - 응답시 매번 새로운 응답 코드 객체를 생성하는 문제 해결

    * Service 계층
    - 유저 회원가입 관련 메소드 이름 변경 (insertUser -> userSignUp)

    * Test
    - duplicateIdCheck 테스트 실패 케이스 추가 및 @DisplayName에 테스트 내용 명시
    - serviceImplTest를 Mockito를 이용하여 단위 테스트로 변경

    * Refactor: 비밀번호 암호화 관련
    - 스프링 security를 없애고 jBcrypt로 대체하여 불필요한 SecurityConfig 설정제거
    - PasswordEncoder 추상화 계층을 도입하여 차후 자유롭게 구현 기술을 바꿀수 있도록 하고 Client는 변화에 영향을 받지 않도록함

Commit:
[#2] 회원가입 기능 구현

- 유저의 서비스 가입을 위한 회원가입 기능추가
- 유저마다 고유한 아이디를 보장하기 위해 아이디 중복체크 구현

Commit:
[#2] 회원가입 기능 구현

- 유저의 서비스 가입을 위한 회원가입 기능추가
- 유저마다 고유한 아이디를 보장하기 위해 아이디 중복체크 구현
- pom.xml 설정 오류 수정

Commit:
[#2] 회원가입 기능 구현

- pom.xml 파일 설정 오타 수정

Commit:
[#2] 피드백 반영

* Controller 계층
- URL관련 웹 API 디자인 지침 반영(/idCheck -> /idcheck)
- 응답시 매번 새로운 응답 코드 객체를 생성하는 문제 해결

* Service 계층
- 유저 회원가입 관련 메소드 이름 변경 (insertUser -> userSignUp)

* Test
- duplicateIdCheck 테스트 실패 케이스 추가 및 @DisplayName에 테스트 내용 명시
- serviceImplTest를 Mockito를 이용하여 단위 테스트로 변경

Commit:
[#2] 피드백 반영

* Refactor: 비밀번호 암호화 관련
- 스프링 security를 없애고 jBcrypt로 대체하여 불필요한 SecurityConfig 설정제거
- PasswordEncoder 추상화 계층을 도입하여 차후 자유롭게 구현 기술을 바꿀수 있도록 하고
  Client는 변화에 영향을 받지 않도록함

f-lab-dev commented on pom.xml:
```diff
@@ -23,18 +23,29 @@
+
+        <dependency>
+            <groupId>org.mindrot</groupId>
+            <artifactId>jbcrypt</artifactId>
```
(팁) 비밀번호 암호화는 이런 외부 라이브러리를 사용 안해도 구현 가능합니다.

`자바 sha256 암호화` 같은 키워드로 검색해보시면 나올겁니다~

msugo1 replied:
넵 알겠습니다 감사합니다.

f-lab-dev commented on src/main/java/me/soo/helloworld/model/Languages.java:
```diff
@@ -0,0 +1,26 @@
+ * 차후 유저의 언어 등록에 활용하기 위한 languages 클래스
+ */
+@Getter @Setter
+@AllArgsConstructor @Builder
```
여기 두 개씩 선언하신거는 리뷰 반영이 안되어있네요~

msugo1 replied:
방금 수정했습니다. 로그인/로그아웃 구현하는 동안 이 부분을 깜빡해버렸네요;;

f-lab-dev commented on src/main/java/me/soo/helloworld/model/Languages.java:
```diff
@@ -0,0 +1,26 @@
+    private final String learningLanguage;
+
+    @NotBlank(message = "배우는 언어의 현재 레벨 정보를 입력해주세요.")
+    private final String learningLanguageLevel;
```
위에 `@Setter`를 선언해두셨는데 final로 만들면 의미가 없는게 아닐까요?

msugo1 replied:
Final의 `immutability` 속성과 setter의 값을 변화시켜주는 속성이 대비되서 쓸모 없어지는게 맞네요. 제거했습니다.

f-lab-dev commented on src/main/java/me/soo/helloworld/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,39 @@
+
+@Service
+@RequiredArgsConstructor
+public class UserServiceImpl implements UserService {
```
`~~Service`-> `~~ServiceImpl` 식의 추상화를 한 이유가 있으실까요?

msugo1 replied:
코드를 구상하면서 다음의 이유로 추상화를 도입하기로 결정해보았습니다.
- 혹시나 모를 차후 변화에 대비
1. `UserServiceImpl` 에 구현된 내용이 크게 변화할 가능성은 낮아보이나 혹시나 변화할 수 있는 상황이 생길 수 있고 이에 Interface를 도입하여  Controller 계층이 추상화된 인터페이스에 의존함으로써 이러한 변화에 영향을 받지 않게 만들고자 의도했습니다.

2. Service 계층의 로직에 약간의 변화를 주어 테스트가 필요할 때 간단히 인터페이스를 구현한 구현체를 만들어주면 간단한 DI만으로 테스트 객체가 기존의 객체를 대신할 수 있을 것이라 생각했습니다. 

보충) **AOP** 적용 관련 (기존의 의도는 아니었지만 학습해가면서 느낀점)
 
  현재 `AOP`의 학습을 진행중인데요. 스프링에서는 AOP를 이용해 부가기능들을 분리하고 이를 다른 적용해줄 때 `프록시` 패턴을 활용합니다. 따라서 차후에 `UserService`에 AOP를 적용할 일이 있으면 Interface를 통해 추상화 해놓은 부분이 큰 역할을 해줄 것이라 생각됩니다.

f-lab-dev replied:
인터페이스가 없어도 프록시는 입힐 수 있습니다~ 테스트도 굳이 인터페이스를 사용하지 않아도 되고요
**변화할 가능성이 얼마나 낮은가**도 고려할 대상인 것 같은데 무조건 이렇게 추상화하는게 좋은걸까요?

msugo1 replied:
여러 게시글을 참고해서 관습적인 추상화와 네이밍 컨벤션의 문제점을 보면서 제가 가진 문제점을 다시 한 번 파악할 수 있었습니다.

- 관습적인 추상화 관련
: Interface의 도입여부를 결정하는 단계에서 제 코드의 방향성에 대한 고려가 부족했던 걸로 생각이 됩니다. 다시금 돌아본 결과 제가 구현하려는 서비스는 사용자들의 유형은 단순히 언어 교환 매칭 상대를 찾는 개개인들이기 때문에 서비스가 확장된다고 하더라도 유저 서비스에 대해 크게 달라질 부분은 없을 것이라 판단했습니다. (혹여나 먼 미래에 자신이 원하는 `Language Teacher/Tutor`를 찾는 기능 등이 도입한다면 사용자의 유형이 조금 더 세분화 될 수 있을 듯 합니다.)

- 네이밍 컨벤션 관련
: 인터페이스의 구현체의 이름 뒤에 단순히 `~Impl`을 붙이는 것은 읽기 좋은 코드를 작성하는데 있어서 전혀 도움을 주지 않기 때문에 이를 지양해야 한다는 것을 깨닫게 되었습니다. 가장 기본적인 `자바 API` 코드들을 살펴봐도 인터페이스를 구현할 때 별 의미없이 뒤에 **Impl**만 붙이는 경우는 없었고, 오히려 이 클래스가 무엇을 나타내는지 명확히 하고자 제대로 된 이름을 붙였습니다. ex) `List -> ArrapyList`  `Map -> HashMap`  등등

## 결론
### `UserService interface`를 빼는 것이 나을 거라 결론을 내리게 되었습니다.
- `UserService`가 미래에 변화할 가능성은 매우 낮을 것으로 예측했습니다. 그렇기 때문에 `UserController`와 `UserService` 사이에 인터페이스를 도입한다고 하더라도 `1:N 관계`에서 인터페이스를 도입했을 때 만큼의 효용은 나올 수 없을 것이며 구현하는 서비스가 커짐에 따라 구조적인 혼란함이 가중될 것이라 생각했습니다. 그러나 혹시 차후에 위에서 언급한 것처럼 단순한 개개인 User가 아니라 Tutor/Teacher 등의 다양한 회원 유형이 추가된다면 그때는 단순히 제가 했던 것처럼 `UserService -> UserServiceImpl` 의 네이밍 컨벤션을 따르는 것이 아니라  `MemberService` -> `UserMemberService/TutorMemberService ~` 등의 좀 더 의미있고 읽는 이름을 붙여주고자 합니다.

f-lab-dev commented on src/main/java/me/soo/helloworld/util/PasswordEncoderBcrypt.java:
```diff
@@ -0,0 +1,17 @@
+
+@Component
+public class PasswordEncoderBcrypt implements PasswordEncoder {
+    @Override
```
취향 차이지만 보통 가독성을 위해 클래스 선언 아래에는 개행을 하나 넣어줍니다~

msugo1 replied:
넵. 공백을 한줄 씩 추가했더니 훨씬 더 보기 편하네요. 앞으로도 코드 작성하면서 계속 어떻게 하면 가독성을 높일지 고민하면서 작성하겠습니다.

Commit:
[#2] 피드백 반영

* Readability
- 가독성 이슈로 annotation 위치 수정: 한줄에 두개 -> 한줄에 한개 씩으로 수정
- 클래스 선언 아래 부분에 개행 한줄씩 추가

* @Setter 관련
- immutable 객체 생성(User, Langauge)을 위한 변수에 final 선언으로 @setter는 무용지물이 되어 제거

Commit:
[#2] 피드백 반영

* 1:1 관계에서 무분별하게 `UserService`가 Interface에 의존하고 있던 부분
  수정
- UserServic interface 삭제, UserServiceImpl 구현체를 UserService로 변경
- 클래스의 변화 가능성이나 확장 가능성이 낮고 차후에도 1:1 구도를 유지할 가능성이 높다고 판단되어 불필요한 관습적적인 추상화와 네이밍 컨벤션`(~ -> ~impl)`을 제거
